x/vote/types/types.go

```diff
@@ -161,81 +175,77 @@ func (p *Poll) getVotingPower(v sdk.ValAddress) int64 {
 	return 0
 }
 
-func (p *Poll) handleRewards() error {
-	majorityVote := p.getMajorityVote()
-
-	for _, vote := range p.GetVotes() {
-		if bytes.Equal(vote.Data.Value, majorityVote.Data.Value) {
-			for _, voter := range vote.Voters {
-				if err := p.rewardPool.ReleaseRewards(voter); err != nil {
-					return err
-				}
-			}
-		} else {
-			for _, voter := range vote.Voters {
-
-				p.logger.Debug("penalizing voter due to incorrect vote",
-					"voter", voter.String(),
-					"poll", p.PollMetadata.Key.String(),
-					"expected", majorityVote.Data.String(),
-					"actual", vote.Data.String())
-				p.rewardPool.ClearRewards(voter)
-			}
-		}
-	}
-
-	return nil
+func (p Poll) isWithinGracePeriod(blockHeight int64) bool {
+	return !p.Is(exported.Expired) &&
+		p.Is(exported.Completed) &&
+		blockHeight <= p.CompletedAt+int64(p.GracePeriod)
 }
 
 // Vote records the given vote
-func (p *Poll) Vote(voter sdk.ValAddress, data codec.ProtoMarshaler) error {
-	if p.Is(exported.NonExistent) {
-		return fmt.Errorf("poll does not exist")
-	}
+func (p *Poll) Vote(voter sdk.ValAddress, blockHeight int64, data codec.ProtoMarshaler) (codec.ProtoMarshaler, bool, error) {
+	switch {
+	case p.Is(exported.NonExistent):
+		return nil, false, fmt.Errorf("poll does not exist")
+	case p.Is(exported.Failed), p.Is(exported.Expired):
+		return nil, false, nil
```

Comment:
So you wanna differentiate what's returned for Failed and Expired?

Reply:
Yeah, it's pretty unexpected to get votes after the poll has expired, which means the validator was really late. However, for failed polls, it could be that it just failed and there were still some votes in the same or next block. There was just disagreement from validators about the events.